Add filters to hide users, move some common code to Temporary users c…

yscik opened the pull request:
Fixes #6386 #6394

### Changes proposed in this Pull Request

* Apply filters to both guest and preview users
* Add more filters to hide these users from the Students and Grading screens too
  * Still show a preview/guest user that has an 'Ungraded' quiz, to support manually graded quizzes
* Always load the admin filters so REST API is filtered too
* Move shared code around to the `Sensei_Temporary_User` class
* Also remove preview users after a week, rename data cleaner class

### Testing instructions

* Create an open access course and take it as a guest student
* Also activate student preview as a teacher for a course and complete a few lessons
* Check that the users and roles are not showing up on the following admin screens:
	* Users → All Users
	* Sensei → Grading
	* Sensei → Students
	* Sensei → Reports
	* Sensei → Groups → [Some group] → Add students
	* On the Sensei screens, also check filtering the pages by course/lesson, status, and searching for users

<!-- Add the following only if there are new/updated actions or filters. Please provide a brief description of what they do and any arguments they may take. Be sure to also add the "Hooks" label to this PR. -->
### New/Updated Hooks

`sensei_learners_query`: Filter the query to list learners on the 'Students' screen

'sensei_check_for_activity': Added a second arguments with the original query arguments.

yscik committed:
Add filters to hide users, move some common code to Temporary users class

yscik committed:
Add more filter options to grading screen

yscik committed:
Remove preview users in data cleaner

yscik committed:
Rename cleaner class

yscik committed:
Fix test issue

Imran92 commented on includes/class-sensei-preview-user.php:
```diff
@@ -387,7 +396,7 @@ public function count_unpublished_lessons( WP_Query $query ) {
 	public function preview_user_always_enrolled( $is_enrolled, $user_id, $course_id ) {
 
-		if ( ! $this->is_preview_user( $user_id ) ) {
+		if ( ! Sensei_Preview_User::is_preview_user( $user_id ) ) {
```
Probably we can change this to `self::is_preview_user` ?

Imran92 commented on includes/class-sensei-preview-user.php:
```diff
@@ -429,7 +438,7 @@ private function set_preview_user( $preview_user_id ) {
 
 		if ( $existing_preview_user ) {
-			$this->delete_preview_user( $existing_preview_user );
+			Sensei_Preview_User::delete_preview_user( $existing_preview_user );
```
Same here

Imran92 commented on includes/class-sensei-preview-user.php:
```diff
@@ -442,7 +451,7 @@ private function set_preview_user( $preview_user_id ) {
 	private function is_preview_user_active() {
 		$user = wp_get_current_user();
-		return $this->is_preview_user( $user );
+		return Sensei_Preview_User::is_preview_user( $user );
```
Here too

Imran92 commented on includes/class-sensei-utils.php:
```diff
@@ -127,17 +127,17 @@ public static function sensei_check_for_activity( $args = array(), $return_comme
+		 * @param array $args Search arguments.
 		 */
-		$comments = apply_filters( 'sensei_check_for_activity', get_comments( $args ) );
+		$comments = apply_filters( 'sensei_check_for_activity', get_comments( $args ), $args );
```
Should we add a `Hooks` label for this one?

yscik replied:
Yup, adding!

Imran92 commented on includes/class-sensei-utils.php:
```diff
@@ -1732,6 +1732,7 @@ public static function update_lesson_status( $user_id, $lesson_id, $status = 'in
 		if ( ! empty( $status ) ) {
 			$args = array(
 				'user_id'   => $user_id,
+				'username'  => get_userdata( $user_id )->user_login ?? null,
```
Maybe I'm missing something, just being curious, are we using this value somewhere?

Imran92 replied:
Got it 👍 it's for the comment_author check ✅

yscik replied:
Yes, just using a similar optimization to avoid having to add a join to the usermeta table.

Imran92 commented on includes/class-sensei-utils.php:
```diff
@@ -1769,11 +1770,12 @@ public static function update_course_status( $user_id, $course_id, $status = 'in
-				'type'    => 'sensei_course_status', /* FIELD SIZE 20 */
-				'action'  => 'update', // Update the existing status...
+				'user_id'  => $user_id,
+				'username' => get_userdata( $user_id )->user_login ?? null,
```
Same for this one

Imran92 commented on includes/class-sensei-temporary-user-cleaner.php:
```diff
@@ -84,37 +84,31 @@ public function clean_inactive_guest_users() {
 	 */
 	private function get_inactive_users() {
+
 		$guest_user_ids = get_users(
```
Should we remove the filter by using `remove_filter( 'pre_user_query', [ Sensei_Temporary_User::class, 'filter_out_temporary_users' ], 11 );` before this query? Otherwise it'll probably not fetch the guest uesrs.

yscik replied:
Sorry about that, forgot to push the commit fixing that test. Should be green now!

Imran92 replied:
Thanks for updating @yscik ! But the cron wasn't still cleaning the users for me. I've checked a bit and found that the `get_users` function here isn't pulling any temporary users. Can you kindly check if it works for you?

For testing- 
- Take a couple of open access courses as Guest user
- Updated the wp_comments for them with old dates
- Install Wp Control plugin
- Find the `sensei_remove_inactive_guest_users` cron event and click on 'Run Now'

yscik committed:
Update static calls

yscik committed:
Fix lint issue

Imran92 reviewed (commented):
Hi @yscik 👋 Another issue I've found is the new temporary users have _001 suffix, it isn't increasing as before.

yscik committed:
Add helper to get unfiltered list of users

yscik committed:
Clean guest/preview users when uninstalling

yscik commented:
Thanks for testing, good catch on those. I added a helper method to get users without the filter, and switched some methods to use that one. Should fix both issues.

yscik committed:
Merge branch 'feature/guest-user' into add/filter-temp-users

Imran92 reviewed (approved):
LGTM! 🚀
